File: platform/sdk/swift/Tests/LayerXSDKTests/ProgramsContractTests.swift
import Foundation
import XCTest
@testable import LayerXSDK

final class ProgramsContractTests: XCTestCase {
    private final class ProgramTransport: PlatformTransport, @unchecked Sendable {
        let response: JSONValue
        init(_ response: JSONValue) { self.response = response }
        func send(_ call: TransportCall) async throws -> JSONValue {
            throw PlatformSDKError(code: .unavailableCapability, retry: .never)
        }
        func sendProgram(_ call: ProgramTransportCall) async throws -> JSONValue { response }
    }

    func testProgramsClientRequiresIndependentNonzeroSequencerPin() throws {
        let client = PlatformClient(transport: ProgramTransport(.emptyObject))
        XCTAssertThrowsError(try ProgramsClient(client: client, sequencerPublicKey: Data(repeating: 0, count: 32)))
        XCTAssertNoThrow(try ProgramsClient(client: client, sequencerPublicKey: Data(repeating: 1, count: 32)))
    }

    func testPendingReceiptMayOmitRetainedBytesButMustBindExpectedActivity() async throws {
        let key = String(repeating: "a", count: 64)
        let activity = Data(repeating: 0x11, count: 32)
        let value: JSONValue = .object([
            "state": .string("unknown"), "activity_id": .string(activity.hexString),
            "idempotency_key": .string(key),
        ])
        let programs = try ProgramsClient(client: PlatformClient(transport: ProgramTransport(value)),
            sequencerPublicKey: Data(repeating: 1, count: 32))
        let pending = try await programs.receipt(idempotencyKey: IdempotencyKey(key), expectedActivityID: activity,
            verificationLevel: "sequencer-signed")
        XCTAssertTrue(pending.isUnknown)
        XCTAssertNil(pending.retainedSignedActivity)
        do {
            _ = try await programs.receipt(idempotencyKey: IdempotencyKey(key),
                expectedActivityID: Data(repeating: 0x12, count: 32), verificationLevel: "sequencer-signed")
            XCTFail("mismatched activity selector was accepted")
        } catch let error as PlatformSDKError {
            XCTAssertEqual(error.code, .verificationFailure)
        }
    }

    func testOperationValueVerificationStatusMatrixIsExact() {
        let achieved: JSONValue = .object(["state": .string("Achieved"), "level": .string("SequencerSigned")])
        let discovery: JSONValue = .object(["state": .string("Unverified"), "level": .string("SequencerSigned"),
            "reason": .string("server_side_receipt_verification_only")])
        let pending: JSONValue = .object(["state": .string("Unverified"), "level": .string("SequencerSigned"),
            "reason": .string("receipt_pending")])
        let unknown: JSONValue = .object(["state": .string("unknown")])
        XCTAssertTrue(AgentHTTPTransport.validVerification("program.discover", value: .emptyObject, status: discovery))
        XCTAssertFalse(AgentHTTPTransport.validVerification("program.discover", value: .emptyObject, status: achieved))
        XCTAssertTrue(AgentHTTPTransport.validVerification("program.receipt", value: unknown, status: pending))
        XCTAssertFalse(AgentHTTPTransport.validVerification("program.receipt", value: unknown, status: achieved))
        XCTAssertTrue(AgentHTTPTransport.validVerification("program.simulate", value: .emptyObject, status: achieved))
        XCTAssertFalse(AgentHTTPTransport.validVerification("program.simulate", value: .emptyObject, status: discovery))
    }

    func testTransferSetV1AndV2ProduceTheSameCanonicalKernelRoot() throws {
        let v1 = transferAuthorization(version: 1)
        let v2 = transferAuthorization(version: 2)
        let rootV1 = try ProgramsWireTestSupport.authorizationRoot(v1)
        let rootV2 = try ProgramsWireTestSupport.authorizationRoot(v2)
        XCTAssertEqual(rootV1, rootV2)
        XCTAssertTrue(rootV1.contains(where: { $0 != 0 }))
        var mutated = v2; mutated[mutated.count - 33] ^= 1
        XCTAssertNotEqual(try ProgramsWireTestSupport.authorizationRoot(mutated), rootV2)
    }

    func testOccupancyV1V2V3AndAggregateBindings() throws {
        let asset = Data(repeating: 0x66, count: 32)
        for version in 1...3 {
            let binding = try ProgramsWireTestSupport.occupancyBinding(emptyOccupancy(version: version), asset: asset)
            XCTAssertEqual(binding.0, UInt128Value(high: 0, low: 0))
            XCTAssertEqual(binding.1, UInt128Value(high: 0, low: 0))
            XCTAssertEqual(binding.2, Data(repeating: 0, count: 32))
        }
        let evidence = chargedOccupancy()
        let binding = try ProgramsWireTestSupport.occupancyBinding(evidence, asset: asset)
        XCTAssertEqual(binding.0, UInt128Value(high: 0, low: 3))
        XCTAssertEqual(binding.1, UInt128Value(high: 0, low: 6))
        XCTAssertTrue(binding.2.contains(where: { $0 != 0 }))
        var mutated = evidence
        let declaredFeeLowByte = Data("LXP/storage-occupancy-settlement/v3\0".utf8).count + 8 + 4 + 7 * 8 + 16 + 15
        mutated[declaredFeeLowByte] ^= 1
        XCTAssertThrowsError(try ProgramsWireTestSupport.occupancyBinding(mutated, asset: asset))
    }

    private func transferAuthorization(version: Int) -> Data {
        let program = Data(repeating: 1, count: 32); let principal = Data(repeating: 2, count: 32)
        let asset = Data(repeating: 4, count: 32); let destination = Data(repeating: 5, count: 32)
        var encoded = Data("LayerX/programs/402LXP/transfer-set/v\(version)\0".utf8)
        encoded.append(program); encoded.append(principal); encoded.append(Data(repeating: 3, count: 32))
        encoded.append(Data(repeating: 0, count: 9))
        var events = Data("LayerX/programs/events/v1\0".utf8); events.append(be32(0))
        encoded.append(be32(UInt32(events.count))); encoded.append(events); encoded.append(be64(0)); encoded.append(be64(1))
        encoded.append(Data(repeating: 0, count: 9))
        if version == 2 { encoded.append(1); encoded.append(principal) }
        encoded.append(asset); encoded.append(destination); encoded.append(be128(7)); encoded.append(program)
        return encoded
    }

    private func emptyOccupancy(version: Int) -> Data {
        var encoded = Data("LXP/storage-occupancy-settlement/v\(version)\0".utf8); encoded.append(be64(1))
        if version > 1 { encoded.append(be32(1)) }
        for value in 1...7 { encoded.append(be64(UInt64(value))) }
        if version == 3 {
            encoded.append(Data(repeating: 0, count: 16 * 4)); encoded.append(be32(0))
        } else {
            encoded.append(Data(repeating: 0, count: 16 * 2)); encoded.append(be64(0))
        }
        return encoded
    }

    private func chargedOccupancy() -> Data {
        let program = Data(repeating: 0x11, count: 32); let payer = Data(repeating: 0x77, count: 32)
        var encoded = Data("LXP/storage-occupancy-settlement/v3\0".utf8); encoded.append(be64(2)); encoded.append(be32(1))
        for value: UInt64 in [0, 0, 0, 0, 0, 0, 2] { encoded.append(be64(value)) }
        encoded.append(be128(3)); encoded.append(be128(6)); encoded.append(be128(6)); encoded.append(be128(0)); encoded.append(be32(1))
        encoded.append(65); encoded.append(program); encoded.append(0); encoded.append(payer)
        encoded.append(payer); encoded.append(program); encoded.append(Data(repeating: 0x88, count: 32))
        encoded.append(be64(1)); encoded.append(be64(2)); encoded.append(be64(3)); encoded.append(be64(3))
        encoded.append(be128(3)); encoded.append(be64(2)); encoded.append(be128(6)); encoded.append(be128(0))
        encoded.append(be128(6)); encoded.append(be128(0)); encoded.append(1); encoded.append(be128(0))
        encoded.append(be64(3)); encoded.append(be64(2)); encoded.append(be128(0)); encoded.append(Data(repeating: 0x99, count: 32))
        return encoded
    }

    private func be32(_ value: UInt32) -> Data { word(value.bigEndian) }
    private func be64(_ value: UInt64) -> Data { word(value.bigEndian) }
    private func be128(_ value: UInt64) -> Data { Data(repeating: 0, count: 8) + be64(value) }
    private func word<T>(_ value: T) -> Data { var copy = value; return withUnsafeBytes(of: &copy) { Data($0) } }
}

private extension Data {
    var hexString: String { map { String(format: "%02x", $0) }.joined() }
}

File: platform/sdk/dotnet/tests/LayerX.Sdk.Tests/ProgramsContractTests.cs
using System.Buffers.Binary;
using System.Numerics;
using System.Reflection;
using System.Text;
using LayerX.Sdk;
using Xunit;

namespace LayerX.Sdk.Tests;

public sealed class ProgramsContractTests
{
    private sealed class ProgramTransport(JsonValue response) : IPlatformTransport
    {
        public Task<JsonValue> SendAsync(TransportCall call, CancellationToken cancellationToken = default) =>
            Task.FromException<JsonValue>(new PlatformSdkException(SdkErrorCode.UnavailableCapability, RetryClass.Never));
        public Task<JsonValue> SendProgramAsync(ProgramTransportCall call, CancellationToken cancellationToken = default) =>
            Task.FromResult(response);
    }

    [Fact]
    public void ProgramsClientRequiresIndependentNonzeroSequencerPin()
    {
        var client = new PlatformClient(new ProgramTransport(JsonValue.EmptyObject));
        Assert.Throws<PlatformSdkException>(() => new ProgramsClient(client, new byte[32]));
        _ = new ProgramsClient(client, Enumerable.Repeat((byte)1, 32).ToArray());
    }

    [Fact]
    public async Task PendingReceiptMayOmitRetainedBytesButMustBindExpectedActivity()
    {
        var key = new string('a', 64); var activity = Enumerable.Repeat((byte)0x11, 32).ToArray();
        var value = JsonValue.Object(new Dictionary<string, JsonValue>
        {
            ["state"] = JsonValue.String("unknown"),
            ["activity_id"] = JsonValue.String(Convert.ToHexString(activity).ToLowerInvariant()),
            ["idempotency_key"] = JsonValue.String(key),
        });
        var programs = new ProgramsClient(new PlatformClient(new ProgramTransport(value)),
            Enumerable.Repeat((byte)1, 32).ToArray());
        var pending = await programs.ReceiptAsync(new IdempotencyKey(key), activity, "sequencer-signed");
        Assert.True(pending.IsUnknown); Assert.Null(pending.RetainedSignedActivity);
        var error = await Assert.ThrowsAsync<PlatformSdkException>(() => programs.ReceiptAsync(new IdempotencyKey(key),
            Enumerable.Repeat((byte)0x12, 32).ToArray(), "sequencer-signed"));
        Assert.Equal(SdkErrorCode.VerificationFailure, error.Code);
    }

    [Fact]
    public void OperationValueVerificationStatusMatrixIsExact()
    {
        var achieved = Status("Achieved", null); var discovery = Status("Unverified", "server_side_receipt_verification_only");
        var pending = Status("Unverified", "receipt_pending");
        var unknown = JsonValue.Object(new Dictionary<string, JsonValue> { ["state"] = JsonValue.String("unknown") });
        Assert.True(TransportStatus("program.discover", JsonValue.EmptyObject, discovery));
        Assert.False(TransportStatus("program.discover", JsonValue.EmptyObject, achieved));
        Assert.True(TransportStatus("program.receipt", unknown, pending));
        Assert.False(TransportStatus("program.receipt", unknown, achieved));
        Assert.True(TransportStatus("program.simulate", JsonValue.EmptyObject, achieved));
        Assert.False(TransportStatus("program.simulate", JsonValue.EmptyObject, discovery));
    }

    [Fact]
    public void TransferSetV1AndV2ProduceTheSameCanonicalKernelRoot()
    {
        var v1 = TransferAuthorization(1); var v2 = TransferAuthorization(2);
        var rootV1 = (byte[])Invoke("DecodeAuthorizationRoot", v1)!;
        var rootV2 = (byte[])Invoke("DecodeAuthorizationRoot", v2)!;
        Assert.Equal(rootV1, rootV2); Assert.Contains(rootV1, value => value != 0);
        var mutated = v2.ToArray(); mutated[^33] ^= 1;
        Assert.NotEqual(rootV2, (byte[])Invoke("DecodeAuthorizationRoot", mutated)!);
    }

    [Fact]
    public void OccupancyV1V2V3AndAggregateBindingsAreExact()
    {
        var asset = Enumerable.Repeat((byte)0x66, 32).ToArray();
        for (var version = 1; version <= 3; version++)
        {
            var binding = Invoke("DecodeOccupancySettlement", EmptyOccupancy(version))!;
            Assert.Equal(BigInteger.Zero, Property<BigInteger>(binding, "ByteBatches"));
            Assert.Equal(BigInteger.Zero, Property<BigInteger>(binding, "FeeUnits"));
            Assert.Equal(new byte[32], (byte[])Invoke("OccupancyTransferRoot", binding, asset)!);
        }
        var evidence = ChargedOccupancy(); var charged = Invoke("DecodeOccupancySettlement", evidence)!;
        Assert.Equal(new BigInteger(3), Property<BigInteger>(charged, "ByteBatches"));
        Assert.Equal(new BigInteger(6), Property<BigInteger>(charged, "FeeUnits"));
        Assert.Contains((byte[])Invoke("OccupancyTransferRoot", charged, asset)!, value => value != 0);
        var mutated = evidence.ToArray();
        var declaredFeeLowByte = Encoding.UTF8.GetByteCount("LXP/storage-occupancy-settlement/v3\0") + 8 + 4 + 7 * 8 + 16 + 15;
        mutated[declaredFeeLowByte] ^= 1;
        var exception = Assert.Throws<TargetInvocationException>(() => Invoke("DecodeOccupancySettlement", mutated));
        Assert.IsType<InvalidDataException>(exception.InnerException);
    }

    private static object? Invoke(string name, params object[] arguments)
    {
        var method = typeof(ProgramsClient).GetMethod(name, BindingFlags.NonPublic | BindingFlags.Static)
            ?? throw new InvalidOperationException($"missing {name}");
        return method.Invoke(null, arguments);
    }

    private static bool TransportStatus(string operation, JsonValue value, JsonValue status)
    {
        var method = typeof(AgentHttpTransport).GetMethod("ValidProgramVerification", BindingFlags.NonPublic | BindingFlags.Static)
            ?? throw new InvalidOperationException("missing ValidProgramVerification");
        return (bool)(method.Invoke(null, [operation, value, status]) ?? false);
    }

    private static JsonValue Status(string state, string? reason)
    {
        var fields = new Dictionary<string, JsonValue>
        {
            ["state"] = JsonValue.String(state), ["level"] = JsonValue.String("SequencerSigned"),
        };
        if (reason is not null) fields["reason"] = JsonValue.String(reason); return JsonValue.Object(fields);
    }

    private static T Property<T>(object value, string name) => (T)(value.GetType().GetProperty(name)?.GetValue(value)
        ?? throw new InvalidOperationException($"missing {name}"));

    private static byte[] TransferAuthorization(int version)
    {
        var program = Repeat(1); var principal = Repeat(2); var asset = Repeat(4); var destination = Repeat(5);
        using var stream = new MemoryStream(); Write(stream, Encoding.UTF8.GetBytes($"LayerX/programs/402LXP/transfer-set/v{version}\0"));
        Write(stream, program); Write(stream, principal); Write(stream, Repeat(3)); Write(stream, new byte[9]);
        using var events = new MemoryStream(); Write(events, Encoding.UTF8.GetBytes("LayerX/programs/events/v1\0")); Write(events, Be(0, 4));
        Write(stream, Be((ulong)events.Length, 4)); Write(stream, events.ToArray()); Write(stream, Be(0, 8)); Write(stream, Be(1, 8));
        Write(stream, new byte[9]); if (version == 2) { stream.WriteByte(1); Write(stream, principal); }
        Write(stream, asset); Write(stream, destination); Write(stream, U128(7)); Write(stream, program); return stream.ToArray();
    }

    private static byte[] EmptyOccupancy(int version)
    {
        using var stream = new MemoryStream(); Write(stream, Encoding.UTF8.GetBytes($"LXP/storage-occupancy-settlement/v{version}\0"));
        Write(stream, Be(1, 8)); if (version > 1) Write(stream, Be(1, 4));
        for (ulong value = 1; value <= 7; value++) Write(stream, Be(value, 8));
        if (version == 3) { Write(stream, new byte[16 * 4]); Write(stream, Be(0, 4)); }
        else { Write(stream, new byte[16 * 2]); Write(stream, Be(0, 8)); }
        return stream.ToArray();
    }

    private static byte[] ChargedOccupancy()
    {
        var program = Repeat(0x11); var payer = Repeat(0x77); using var stream = new MemoryStream();
        Write(stream, Encoding.UTF8.GetBytes("LXP/storage-occupancy-settlement/v3\0")); Write(stream, Be(2, 8)); Write(stream, Be(1, 4));
        foreach (ulong value in new ulong[] { 0, 0, 0, 0, 0, 0, 2 }) Write(stream, Be(value, 8));
        Write(stream, U128(3)); Write(stream, U128(6)); Write(stream, U128(6)); Write(stream, U128(0)); Write(stream, Be(1, 4));
        stream.WriteByte(65); Write(stream, program); stream.WriteByte(0); Write(stream, payer);
        Write(stream, payer); Write(stream, program); Write(stream, Repeat(0x88));
        Write(stream, Be(1, 8)); Write(stream, Be(2, 8)); Write(stream, Be(3, 8)); Write(stream, Be(3, 8));
        Write(stream, U128(3)); Write(stream, Be(2, 8)); Write(stream, U128(6)); Write(stream, U128(0));
        Write(stream, U128(6)); Write(stream, U128(0)); stream.WriteByte(1); Write(stream, U128(0));
        Write(stream, Be(3, 8)); Write(stream, Be(2, 8)); Write(stream, U128(0)); Write(stream, Repeat(0x99));
        return stream.ToArray();
    }

    private static byte[] Repeat(byte value) => Enumerable.Repeat(value, 32).ToArray();
    private static byte[] U128(ulong value) => new byte[8].Concat(Be(value, 8)).ToArray();
    private static byte[] Be(ulong value, int length)
    {
        var encoded = new byte[8]; BinaryPrimitives.WriteUInt64BigEndian(encoded, value); return encoded[(8 - length)..];
    }
    private static void Write(Stream stream, byte[] value) => stream.Write(value);
}
